### DISCUSSION TITLE
Why pointer vs. value in the generated code?

### DISCUSSION THREAD

**Discussion by burdiyan** [stephenafamo/bob#43] [Q&A] [answered]
The documentation talks about preferring pointers vs. values in model slices, but doesn’t explain why. Any particular reason for this choice? 

https://bob.stephenafamo.com/docs/code-generation/usage

```
// JetSlice is an alias for a slice of pointers to Jet.
// This should almost always be used instead of []Jet.
type JetSlice []*Jet

var JetsTable = psql.NewTablex[*Jet, JetSlice, *JetSetter]("", "jets")
```

**Comment by stephenafamo** [comment, marked as answer]
Actually, I should change the comment to be `instead of []*Jet`

The recommendation is not about pointers vs values, but more about using the defined type `JetSlice` over `[]*Jet`. The reason is that there are some convenient methods defined on `JetSlice` that would not be available when using the plain slice.

**Reply by burdiyan** [reply]
Oh, I see your point now. The recommendation is to use the named `JetSlice` type vs. plain slice.

Although, still seems like the generated code encourages using pointers and not values (looking at the last line in the first comment):

```
var JetsTable = psql.NewTablex[*Jet, JetSlice, *JetSetter]("", "jets")
```

Why the table models choose to return `*Jet` or `[]*Jet`, and accept `*JetSetter`, instead of using non-pointer values? My concern is especially with slices, where each slice item would end up allocated on the heap when using pointers.

**Reply by stephenafamo** [reply]
The methods are all defined on pointers, and this is particularly useful for all the methods that modify the object itself (e.g. loading related objects). For this reason, it is better to use a pointer to the model.

`JetSlice` is in a similar situation. Some of the methods on the slice modify the structs within, so it has to contain pointers and not just values.  
Since we deal with pointers to the models themselves, the slice also has to be a slice of pointers.
